### ISSUE TITLE
false positive on Literal string type to string comparison

### ISSUE THREAD

**Issue by Sage0614** [microsoft/pyright#5731]
**Describe the bug**
pyright cannot correctly infer Literal string assigned to variable

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

```python
from typing import Literal,TypeAlias

TaskTypes: TypeAlias = Literal["train", "validation", "test"]

class Fitter:
    def __init__(
        self,
        task: TaskTypes,
    ):
        self.task = task
    
    @staticmethod
    def fit(task: TaskTypes) -> None:
        print(f"fitting {task} task")
        
    def run(self) -> None:
        self.fit(self.task)
```

error message:
`
Argument of type "str" cannot be assigned to parameter "task" of type "TaskTypes" in function "fit"
  Type "str" cannot be assigned to type "TaskTypes"
    "str" cannot be assigned to type "Literal['train']"
    "str" cannot be assigned to type "Literal['validation']"
    "str" cannot be assigned to type "Literal['test']"
`

I run mypy to check the above code as well, mypy doesn't complain about this.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

Pylance version:  v2023.8.20

**Comment by erictraut** [comment]
This is intended behavior. You haven't provided a type declaration for the instance variable `task`, so its type is inferred. Pyright's normal inference rules do not retain literals. This matches normal expectations. For example, if you were to assign `self.task = 1`, you probably wouldn't expect `task` to be restricted to `Literal[1]`. Instead, you'd expect its type to be inferred as `int`. It's the same in your example.

If you want the type of `task` to be restricted to a literal value or a union of literals, you need to declare it as such.

```python
class Fitter:
    def __init__(self, task: TaskTypes):
        self.task: TaskTypes = task
```

For more details, refer to [the documentation](https://microsoft.github.io/pyright/#/type-inference?id=literals).
